livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,423 @@
+        api_key: str,
+        http_session: aiohttp.ClientSession,
+        sample_rate: int = 16000,
+        num_channels: int = 1,
+        max_retry: int = 32,
+        token_expires_in: int = 3600,
+        buffer_size_seconds: float = 0.2,
+        end_utterance_silence_threshold: int = 1000,
+    ) -> None:
+        super().__init__()
+
+        self._opts = opts
+        self._sample_rate = sample_rate
+        self._num_channels = num_channels
+        self._api_key = api_key
+        self._speaking = False
+        self._session = http_session
+        self._queue = asyncio.Queue[Union[rtc.AudioFrame, str]]()
+        self._event_queue = asyncio.Queue[Optional[stt.SpeechEvent]]()
+        self._closed = False
+        self._main_task = asyncio.create_task(self._run(max_retry))
+
+        if self._num_channels != 1:
+            raise ValueError(f"AssemblyAI only supports mono audio, but a `num_channels` of {self._num_channels} was provided")
+
+        # keep a list of final transcripts to combine them inside the END_OF_SPEECH event
+        self._final_events: List[stt.SpeechEvent] = []
+
+        self._token_expires_in = token_expires_in
+        self._token = None
+
+        self._END_UTTERANCE_SILENCE_THRESHOLD_MSG: str = json.dumps({"end_utterance_silence_threshold": end_utterance_silence_threshold })
+        
+        # Buffer to collect frames until we have 100ms worth of audio
+        self._buffer_size_seconds = buffer_size_seconds
+
+    def push_frame(self, frame: rtc.AudioFrame) -> None:
+        if self._closed:
+            raise ValueError("cannot push frame to closed stream")
+
+        self._queue.put_nowait(frame)
+
+    async def aclose(self, *, wait: bool = True) -> None:
+        self._closed = True
+        self._queue.put_nowait(SpeechStream._CLOSE_MSG)
+
+        if not wait:
+            self._main_task.cancel()
+
+        with suppress(asyncio.CancelledError):
+            await self._main_task
+
+        await self._session.close()
+
+    async def _get_temporary_token(self):
+        url = "https://api.assemblyai.com/v2/realtime/token"
+        headers = {
+            "Authorization": self._api_key,
+            "Content-Type": "application/json",
+        }
+        data = json.dumps(
+            {
+                "expires_in": self._token_expires_in,
+            },
+        )
+        res = await self._session.post(url, headers=headers, data=data)
+        res_json = await res.json()
+        self._token = res_json["token"]
+        if self._token is None:
+            raise ValueError("Failed to get temporary token")
+        self._token_expires_at = time.time() + self._token_expires_in
+
+    async def _run(self, max_retry: int) -> None:
+        """
+        Run a single websocket connection to AssemblyAI and make sure to reconnect
+        when something went wrong.
+        """
+        if self._token is None or time.time() > self._token_expires_at:
+            await self._get_temporary_token()
```

Comment:
should we move this into the while loop in case the token happens to expire during a retry?

Reply:
Yes, I think we should just use the API key